### PR TITLE
WIP Add comments to solvers.jl

### PR DESCRIPTION
**Pull request by IainNZ**
I hadn't looked at this stuff in a while, so while I review it I thought I'd add comments for the sake of future generations.

### CODE REVIEW

File: src/solvers.jl [hunk 1]
```diff
@@ -88,20 +120,31 @@ function solve(m::Model; suppress_warnings=false, ignore_solve_hook=(m.solvehook
...
             m.colVal = colVal
         end
     end
 
+    # The MathProgBase interface defines a conic problem to always be
+    # a minimization problem, so we need to flip the objective before
+    # reporting it to the user
     if traits.conic && m.objSense == :Max
         m.objVal *= -1
         scale!(m.linconstrDuals, -1)
```
Comment (IainNZ) [inline src/solvers.jl]: @mlubin this is a NOOP right because we don't support duals for conic problems, right?

Comment (mlubin) [inline src/solvers.jl]: Yep

File: src/solvers.jl [hunk 2]
```diff
@@ -2,20 +2,31 @@
...
-    lin::Bool
-    qp ::Bool
-    qc ::Bool
-    nlp::Bool
-    soc::Bool
-    sdp::Bool
-    sos::Bool
-    conic::Bool
+    int::Bool  # has integer variables
+    lin::Bool  # has only linear constraints
```
Comment (mlubin) [inline src/solvers.jl]: and linear objective?

File: src/solvers.jl [hunk 3]
```diff
@@ -2,20 +2,31 @@
...
+    qp ::Bool  # has a quadratic objective function
+    qc ::Bool  # has a quadratic constraint
+    nlp::Bool  # has general nonlinear objective or constraints
+    soc::Bool  # has a second-order cone constraint
+    sdp::Bool  # has an SDP constraint (or SDP variable bounds)
+    sos::Bool  # has an SOS constraint
+    conic::Bool  # has an SDP or SOS constraint
```
Comment (mlubin) [inline src/solvers.jl]: SDP or SOC

File: src/solvers.jl [hunk 4]
```diff
@@ -64,7 +107,12 @@ function solve(m::Model; suppress_warnings=false, ignore_solve_hook=(m.solvehook
             end
         end
     else
+        # Problem was not solve to optimality, attempt to extract useful
```
Comment (mlubin) [inline src/solvers.jl]: solved*

**Comment by IainNZ** [comment]
Perf changes for the `verify_ownership` change.
Times both on 0.4, with JuMP compiled. On this branch, so using `verify_ownership` throughout.
```
BEFORE
PMEDIAN BUILD MIN=0.539766973  MED=0.829337445
PMEDIAN INTRN MIN=4.777901945  MED=4.856417003
CONT5 BUILD   MIN=0.18575696  MED=0.367095097
CONT5 INTRN   MIN=0.511391816  MED=0.535300503
```
```
AFTER
PMEDIAN BUILD MIN=0.574436618  MED=0.835707421
PMEDIAN INTRN MIN=4.345520664  MED=4.43239901
CONT5 BUILD   MIN=0.183608798  MED=0.310254336
CONT5 INTRN   MIN=0.472166626  MED=0.519696314
```
So pretty marginal, but something?

**Comment by IainNZ** [comment]
I'll probably keep going with this, because its turning up some real stuff, but this PR was going to get a bit long (even with clean discrete commits)

File: src/JuMP.jl [hunk 5]
```diff
@@ -380,7 +380,13 @@ Base.zero(::Variable) = zero(Variable)
 Base.one(::Type{Variable}) = AffExpr(Variable[],Float64[],1.0)
 Base.one(::Variable) = one(Variable)
 
-verify_ownership(m::Model, vec::Vector{Variable}) = all(v->isequal(v.m,m), vec)
+function verify_ownership(m::Model, vec::Vector{Variable})
+    n = length(vec)
+    @inbounds for i in 1:n
+        !isequal(vec[i].m, m) && return false
```
Comment (mlubin) [inline src/JuMP.jl]: why ``!isequal`` instead of ``!===``?

Comment (IainNZ) [inline src/JuMP.jl]: Shrug, @joehuchette ? I just copied what was there

**Pull request merged**